src/cargo/ops/cargo_compile.rs

```diff
@@ -720,7 +738,12 @@ fn generate_targets<'a>(
         let profile =
             bcx.profiles
                 .get_profile(pkg.package_id(), ws.is_member(pkg), unit_for, target_mode);
-        let features = resolve.features_sorted(pkg.package_id());
+
+        let features = Vec::from(resolved_features.activated_features(
+            pkg.package_id(),
+            DepKind::Normal,
+            kind,
+        ));
```

Comment:
Root units are not dependencies of anything. Their features come from the command-line options.

For test targets, nothing can depend on a test, so there are no edges going from dev-dependency to the test. Thus there shouldn't be any dev-dependency features injected into the test.

proc_macros are not supported (because we don't know they exist at resolve time).

I've been trying to think if there's some way to use `[patch]` or cyclical dev-dependencies to break these assumptions, but I haven't been able to come up with anything.